src/libos/crates/async-io/src/util/channel.rs

```diff
@@ -158,9 +181,83 @@ impl Producer {
             peer_end.pollee().add_events(Events::IN);
         }
     }
+
+    pub fn shutdown(&self) {
+        self.common.producer.shutdown()
+    }
+
+    pub fn is_shutdown(&self) -> bool {
+        self.common.producer.is_shutdown()
+    }
+
+    pub fn status_flags(&self) -> StatusFlags {
+        self.this_end().flags.load(Ordering::Relaxed)
+    }
+
+    pub fn set_status_flags(&self, new_status: StatusFlags) -> Result<()> {
+        check_status_flags(new_status)?;
+        self.this_end().flags.store(new_status, Ordering::Relaxed);
+        Ok(())
+    }
+
+    pub fn is_nonblocking(&self) -> bool {
+        self.status_flags().contains(StatusFlags::O_NONBLOCK)
+    }
+
+    pub fn pollee(&self) -> &Pollee {
+        self.common.producer.pollee()
+    }
+}
+
+impl Producer<u8> {
+    pub async fn writev_async(&self, bufs: &[&[u8]]) -> Result<usize> {
+        let total_len: usize = bufs.iter().map(|buf| buf.len()).sum();
+        if total_len == 0 {
+            return Ok(0);
+        }
+
+        let mut poller = None;
+        loop {
+            // Attempt to write
+            let ret = self.try_writev(bufs);
+            if !ret.has_errno(EAGAIN) {
+                return ret;
+            }
+
+            if self.is_nonblocking() {
+                return_errno!(EAGAIN, "buffer is full");
+            }
+
+            // Wait for interesting events by polling
+            if poller.is_none() {
+                poller = Some(Poller::new());
+            }
+            let mask = Events::OUT;
+            let events = self.pollee().poll(mask, poller.as_mut());
+            if events.is_empty() {
+                poller.as_ref().unwrap().wait().await;
+            }
+        }
+    }
+
+    fn try_writev(&self, bufs: &[&[u8]]) -> Result<usize> {
+        let mut total_nbytes = 0;
+        for buf in bufs {
+            let ret = self.write(buf);
+            if let Ok(nbytes) = ret {
+                total_nbytes += nbytes;
+            }
```

Comment:
Should we return the error here?

Reply:
Refine the logic here. Thanks for your review.